### PR TITLE
[android][web-browser] Fix `openAuthSessionAsync` stating browser is already open

### PR DESCRIPTION
**Pull request by sumnerwarren**
# Why

As shown in expo/expo#12045, it is possible for `openBrowserAsync` to fail before the browser opens. When this was prompted by a call to `openAuthSessionAsync` it should be possible to call `openAuthSessionAsync` again without seeing the error `WebBrowser is already open, only one can be open at a time`. An example where this may prove useful is if the user leaves the app to change their default browser settings (or which browsers are installed) and returns to attempt authentication again. Without this fix, `openAuthSessionAsync` will think a browser is open when it is not.

# How

I fixed this bug by improving the internal state check for whether a browser was open and returning the appropriate error in both cases.

# Test Plan

Added a unit test to make sure `_onWebBrowserCloseAndroid` is cleared properly when the browser fails to open to do intent resolution issues. The original issue and fix can be observed in a Pixel 4 API 30 emulator with the DuckDuckGo app set as the default browser. Before this change all `openAuthSessionAsync` attempts except the first would erroneously state a browser was already open. After this change all `openAuthSessionAsync` attempts should throw the `No matching activity!` error.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-web-browser/src/WebBrowser.ts [hunk 1]
```diff
@@ -355,7 +355,17 @@ async function _openBrowserAndWaitAndroidAsync(
   });
 
   let result: WebBrowserResult = { type: WebBrowserResultType.CANCEL };
-  const { type } = await openBrowserAsync(startUrl, browserParams);
+  let type: string | null = null;
+
+  try {
+    ({ type } = await openBrowserAsync(startUrl, browserParams));
+  } catch (e) {
+    if (e.message === 'No matching activity!') {
```
Comment (bbarthec) [inline packages/expo-web-browser/src/WebBrowser.ts:363]: I believe we should clear the state no matter the type of error 🤔

**Review by bbarthec — CHANGES_REQUESTED** [review]
@sumnerwarren, thank you very much for finding the culprit for this erroneous scenario 👏 
I think the approach you're proposing is fine, but I would say that we can cleanup the state no matter the type of error 🧹 
Let's remove the `if` condition and we're almost ready to merge it 👍

**Comment by sumnerwarren** [comment]
> Let's remove the `if` condition and we're almost ready to merge it 👍

All set! I updated the test to demonstrate multiple error types as well.

File: packages/expo-web-browser/src/__tests__/WebBrowser-test.native.ts [hunk 2]
```diff
@@ -31,3 +31,27 @@ it(`dismissBrowser returns nothing`, () => {
...
+  ExpoWebBrowser.openBrowserAsync.mockRejectedValueOnce(new Error('No matching activity!'));
+  ExpoWebBrowser.openBrowserAsync.mockRejectedValueOnce(new Error('Current activity not found!'));
+  ExpoWebBrowser.openBrowserAsync.mockRejectedValueOnce(new Error('No package manager!'));
+  const pageUrl = 'http://expo.io';
+  const redirectUrl = 'exp://expo.io';
+  try {
+    await WebBrowser.openAuthSessionAsync(pageUrl, redirectUrl);
+  } catch (e) {
+    expect(e.message).toBe('No matching activity!');
+  }
```
Comment (ide) [inline packages/expo-web-browser/src/__tests__/WebBrowser-test.native.ts:45]:
```suggestion
  await expect(WebBrowser.openAuthSessionAsync(pageUrl, redirectUrl))
    .rejects
    .toThrowError('No matching activity!');
```
etc..

File: packages/expo-web-browser/src/__tests__/WebBrowser-test.native.ts [hunk 3]
```diff
@@ -31,3 +31,27 @@ it(`dismissBrowser returns nothing`, () => {
...
+    await WebBrowser.openAuthSessionAsync(pageUrl, redirectUrl);
+  } catch (e) {
+    expect(e.message).toBe('Current activity not found!');
+  }
+  try {
+    await WebBrowser.openAuthSessionAsync(pageUrl, redirectUrl);
+  } catch (e) {
+    expect(e.message).toBe('No package manager!');
+  }
+  expect.assertions(3);
```
Comment (ide) [inline packages/expo-web-browser/src/__tests__/WebBrowser-test.native.ts:56]: then don't need this
```suggestion
```

**Review by bbarthec — COMMENTED** [review]
@sumnerwarren thank you for taking care of the suggestions 👏 
Make sure you're replied to all of them and we're ready to merge it (if only CI is green) 👍

**Pull request merged**